face/unix-stream-transport.go

```diff
@@ -117,20 +117,32 @@ func (t *UnixStreamTransport) runReceive() {
 		}
 
 		// Determine whether valid packet received
-		for startPos > 0 {
-			_, _, tlvSize, err := tlv.DecodeTypeLength(recvBuf)
+		tlvPos := 0
+		for {
+			if tlvPos >= startPos {
+				startPos = 0
+				break
+			}
+
+			_, _, tlvSize, err := tlv.DecodeTypeLength(recvBuf[tlvPos:])
 			if err != nil {
 				core.LogInfo(t, "Unable to process received packet: ", err)
 				startPos = 0
 				break
-			} else if startPos >= tlvSize {
+			} else if startPos >= tlvPos+tlvSize {
 				// Packet was successfully received, send up to link service
-				t.linkService.handleIncomingFrame(recvBuf[:tlvSize])
-				startPos -= tlvSize
-				if startPos > 0 {
-					copy(recvBuf, recvBuf[tlvSize:])
-				}
+				t.linkService.handleIncomingFrame(recvBuf[tlvPos : tlvPos+tlvSize])
+				tlvPos += tlvSize
 			} else {
+				if tlvPos > 0 {
+					if startPos > tlvPos {
+						// Move remaining data to beginning of buffer
+						copy(recvBuf, recvBuf[tlvPos:startPos])
```

Comment:
Now we have another setting. Suppose we have received 100B consisting of a 90B packet and a 10B fragment of another 15B packet.
After the first round, the program moves here, copies the first 10B to recvBuf, set `startPos = 10`, and goes on looping.
Now, suppose the remaining 5B is received. By L120, `tlvPos = 0`, so the program skips the condition at L122 and moves to L127, which tries to decode the packet from its 11th byte (i.e. in the middle).
Is this supposed behaviour?

Reply:
OK. Sorry I missed the buffer copying here. This code is different from what I usually write.